tests/test_cli.py

```diff
@@ -0,0 +1,102 @@
+"""Tests for the CLI."""
+import os
+import pathlib
+import sys
+from unittest import mock
+
+import pytest
+
+from open_alchemy import cli
+from open_alchemy import exceptions
+
+
+@pytest.mark.parametrize(
+    "specfile",
+    [
+        pytest.param("spec.json", id="valid json specfile"),
+        pytest.param("spec.yml", id="valid yml specfile"),
+        pytest.param("spec.yaml", id="valid yaml specfile"),
+    ],
+)
+@pytest.mark.cli
+def test_validate_specfile_valid(tmp_path, specfile):
+    """
+    GIVEN a valid specification file
+    WHEN validate_specfile is called
+    THEN nothing happens
+    """
+    spec = tmp_path / specfile
+    spec.write_text("")
+    cli.validate_specfile(spec)
+
+
+@pytest.mark.cli
+def test_validate_specfile_invalid_extension(tmp_path):
+    """
+    GIVEN a invalid specification file extension
+    WHEN validate_specfile is called
+    THEN a ValueError exception is raised
+    """
+    spec = tmp_path / "specfile.txt"
+    spec.write_text("")
+    with pytest.raises(exceptions.BaseError):
+        cli.validate_specfile(spec)
+
+
+@pytest.mark.cli
+def test_validate_specfile_does_not_exist(tmp_path):
+    """
+    GIVEN a specification file which does not exist
+    WHEN validate_specfile is called
+    THEN a ValueError exception is raised
+    """
+    spec = tmp_path / "specfile.txt"
+    with pytest.raises(exceptions.BaseError):
+        cli.validate_specfile(spec)
+
+
+@pytest.mark.cli
+def test_build_application_parser():
+    """
+    GIVEN an argument on the command line
+    WHEN the application parser is built
+    THEN arguments are returned
+    """
+    sys.argv = ["--help"]
+    args = cli.build_application_parser()
+    assert args
+
+
+@pytest.mark.cli
+def test_build(mocker):
+    """
+    GIVEN arguments from the parser
+    WHEN they are passed to the build() function
+    THEN the function execute
+    """
+    mocker.patch("open_alchemy.cli.validate_specfile")
+    mocker.patch("open_alchemy.cli.build_json")
+    args = mock.MagicMock()
+    args.specfile = pathlib.Path("spec.json")
+    cli.build(args)
+
+
+@pytest.mark.parametrize(
+    "command",
+    [
+        pytest.param(
+            "build examples/simple/example-spec.yml simple .",
+            id="cli build all",
+        ),
+    ],
+)
+@pytest.mark.cli
+def test_main(tmp_path, command):
+    """
+    GIVEN CLI options are set
+    WHEN the CLI is called
+    THEN the program runs
+    """
+    os.chdir(tmp_path)
```

Comment:
Does this impact the state for other tests or is this reset after the test executes?

Reply:
This is a good question. I would say no because `tmp_path` is temporary and the tests are scoped, but I'll definitely research that and provide a more "verified" answer.

Reply:
I added `print(os.getcwd()` in all cli tests, then ran `pytest -x -s -m cli` and it gave this output:

```
tests/test_cli.py::test_build_application_parser /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_invalid_extension /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid yml specfile] /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_does_not_exist /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid json specfile] /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_build /Users/rgreinhofer/projects/rgreinho/OpenAlchemy
PASSED
tests/test_cli.py::test_main[cli build all] /private/var/folders/6g/038mjwy91b7_m7zvb9_r_hjw0000gp/T/pytest-of-rgreinhofer/pytest-41/test_main_cli_build_all_0
PASSED
tests/test_cli.py::test_validate_specfile_valid[valid yaml specfile] /private/var/folders/6g/038mjwy91b7_m7zvb9_r_hjw0000gp/T/pytest-of-rgreinhofer/pytest-41/test_main_cli_build_all_0
PASSED
```

I believe this proves that the `os.chdir()` function only impact the current test. 👍

Reply:
Good stuff :smile: